pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerMemoryLeakTest.java

```diff
@@ -0,0 +1,367 @@
+                    public void close() {
+                        if (method.equals("close")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+                    }
+
+                    @Override
+                    public boolean eligible(Message message) {
+                        if (method.equals("eligible")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+                        return false;
+                    }
+
+                    @Override
+                    public Message beforeSend(Producer producer, Message message) {
+                        if (method.equals("beforeSend")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+                        return message;
+                    }
+
+                    @Override
+                    public void onSendAcknowledgement(Producer producer, Message message, MessageId msgId,
+                                                      Throwable exception) {
+                        if (method.equals("onSendAcknowledgement")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+
+                    }
+                }).create();
+
+        MsgPayloadTouchableMessageBuilder<String> msgBuilder = newMessage(producer);
+        try {
+            msgBuilder.value("msg-1").sendAsync().get(3, TimeUnit.SECONDS);
+            // It may throw error.
+        } catch (Exception ex) {
+            assertTrue(ex.getMessage().contains("Mocked"));
+        }
+
+        // Verify: message payload has been released.
+        producer.close();
+        assertEquals(msgBuilder.payload.refCnt(), 1);
+
+        // cleanup.
+        msgBuilder.release();
+        assertEquals(msgBuilder.payload.refCnt(), 0);
+        admin.topics().delete(topicName);
+    }
+
+    private <T> MsgPayloadTouchableMessageBuilder<T> newMessage(ProducerImpl<T> producer){
+        Schema<T> schema = getInternalState(producer, "schema");
+        return new MsgPayloadTouchableMessageBuilder<T>(producer, schema);
+    }
+
+    private static class MsgPayloadTouchableMessageBuilder<T> extends TypedMessageBuilderImpl {
+
+        public volatile ByteBuf payload;
+
+        private volatile Function<MessageImpl<T>, MessageImpl<T>> msgMocker;
+
+        public <T> MsgPayloadTouchableMessageBuilder(ProducerBase producer, Schema<T> schema) {
+            super(producer, schema);
+        }
+        public void setMsgMocker(Function<MessageImpl<T>, MessageImpl<T>> msgMocker) {
+            this.msgMocker = msgMocker;
+        }
+
+        @Override
+        public Message<T> getMessage() {
+            beforeSend();
+            MessageImpl<T> msg = MessageImpl.create(msgMetadata, content, schema,
+                    producer != null ? producer.getTopic() : null);
+            payload = getInternalState(msg, "payload");
```

Comment:
Whenever possible, don't use reflection unless it's hard to change the visibility of the private field. In this case, `MessageImpl` already provides a `getPayload` method visible for testing. Even it does not provide, it's better to add one rather than using reflection.

Reply:
Solved